app/src/main/java/com/jerboa/ui/components/blocks/BlocksActivity.kt

```diff
@@ -0,0 +1,242 @@
+import com.jerboa.ui.theme.Title
+import it.vercruysse.lemmyapi.v0x19.datatypes.CommunityId
+import it.vercruysse.lemmyapi.v0x19.datatypes.Instance
+import it.vercruysse.lemmyapi.v0x19.datatypes.PersonId
+
+@Composable
+private fun BlockedElementListItem(
+    apiState: ApiState<*>,
+    icon: String?,
+    name: String,
+    onUnblock: () -> Unit,
+    onSuccessfulUnblock: () -> Unit,
+) {
+    Row(verticalAlignment = Alignment.CenterVertically) {
+        icon?.let {
+            CircularIcon(
+                icon = it,
+                contentDescription = null,
+                size = ICON_SIZE,
+            )
+            Spacer(modifier = Modifier.padding(horizontal = SMALL_PADDING))
+        }
+        Text(name, modifier = Modifier.weight(1f))
+        TextButton(onClick = onUnblock, colors = ButtonDefaults.buttonColors(Color.Transparent)) {
+            when (apiState) {
+                ApiState.Loading -> CircularProgressIndicator(
+                    modifier = Modifier.size(SMALL_ICON_SIZE),
+                    color = MaterialTheme.colorScheme.secondary,
+                )
+
+                is ApiState.Success -> onSuccessfulUnblock()
+                else -> Icon(
+                    imageVector = Icons.Rounded.Close,
+                    modifier = Modifier.size(SMALL_ICON_SIZE),
+                    tint = MaterialTheme.colorScheme.error,
+                    contentDescription = null,
+                )
+            }
+        }
+    }
+}
+
+@Preview(
+    showBackground = true,
+    widthDp = 360,
+)
+@Composable
+private fun BlockedElementListItemPreview() {
+    BlockedElementListItem(
+        apiState = ApiState.Empty,
+        icon = "",
+        name = "Element name",
+        onUnblock = { },
+        onSuccessfulUnblock = { },
+    )
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun BlocksActivity(
+    siteViewModel: SiteViewModel,
+    onBack: () -> Unit,
+) {
+    val context = LocalContext.current
+    val snackbarHostState = remember { SnackbarHostState() }
+    val listState = rememberLazyListState()
+
+    LaunchedEffect(Unit) { siteViewModel.getSite() }
+
+    Scaffold(
+        snackbarHost = { JerboaSnackbarHost(snackbarHostState) },
+        topBar = {
+            SimpleTopAppBar(text = stringResource(R.string.blocks), onClickBack = onBack)
+        },
+        content = { padding ->
+            when (val siteRes = siteViewModel.siteRes) {
+                ApiState.Empty -> ApiEmptyText()
+                ApiState.Loading -> LoadingBar(padding)
+
+                is ApiState.Failure -> ApiErrorText(siteRes.msg)
+
+                is ApiState.Success -> {
+                    val personBlocks = siteRes.data.my_user?.person_blocks?.toMutableStateList()
+                    val communityBlocks =
+                        siteRes.data.my_user?.community_blocks?.toMutableStateList()
+                    val instanceBlocks = siteRes.data.my_user?.instance_blocks?.toMutableStateList()
```

Comment:
Needs to be in the viewmodel. 

![image](https://github.com/LemmyNet/jerboa/assets/67873169/f5ca4952-8844-45a5-a2c7-a2430af15404)

The `onSuccessfulUnblock` needs to be removed from the UI. Instead the ViewModel updates the SnapShotList (Block event). The UI will react to this change and update. *Dont use a list, use mutableStateList and only expose SnapShotList (Its what I am doing with changes to the feed, atm we create a new list there constantly but causes the entire feed to recompose each time)*

Reply:
This would be a different viewmodel? The ones I created so far are viewmodels for each list element.

Reply:
Yes that was what I had in mind. But if you have a solution that works with viewmodel per element you can do that too. But I think it would be a lot simpler if you used just one viewmodel.

Reply:
Agree, a BlocksViewModel sounda good. I'll see later what I can do about the unique viewmodel that represents a list item